File: JavaScript/7js 3/squaresNeeded.js
/*
|--------------------------------------------------------------------------
| The wheat/rice and chessboard problem
|--------------------------------------------------------------------------
|
*/

function squaresNeeded(grains) {
   return grains ? grains.toString(2).length : 0;
}


function squaresNeeded(grains) {
   return Math.ceil(Math.log2(grains + 1));
}


function squaresNeeded(grains) {
   return grains < 1 ? 0 : 1 + squaresNeeded(Math.floor(grains / 2)); 
}


console.log(squaresNeeded(0), 0);
console.log(squaresNeeded(1), 1);
console.log(squaresNeeded(2), 2);
console.log(squaresNeeded(3), 2);
console.log(squaresNeeded(4), 3);

File: Python/7py 3/squares_needed.py
"""
|--------------------------------------------------------------------------
| The wheat/rice and chessboard problem
|--------------------------------------------------------------------------
|
"""

def squares_needed(grains):
   return len(bin(grains)[2:]) if grains else 0


def squares_needed(grains):
   return grains.bit_length()


def squares_needed(grains):
   return 0 if grains<1 else 1+squares_needed(grains//2)


print(squares_needed(0), 0)
print(squares_needed(1), 1)
print(squares_needed(2), 2)
print(squares_needed(3), 2)
print(squares_needed(4), 3)